ROS2/ROS_workspace/src/cli_app/src/main.cpp

```diff
@@ -30,35 +49,58 @@ int main(int argc, char* argv[]) {
             printCLICommands();
             return 0;
         } else if (arg == "launch_gui") {
-            std::thread guiThread(launchGUI, false);
+            std::thread guiThread(launchGUI, controller_input_publisher, false);
             guiThread.detach();
 
         }else if (arg == "launch_controller") {
-            std::thread controlThread(launchGUI, true);
+            std::thread controlThread(launchGUI, controller_input_publisher, true);
             controlThread.detach();
 
         }
         
     }
 
-    std::string input;
-
     printColoredAsciiImage();
     printIntroduction();
 
     const std::chrono::milliseconds loop_duration(10); // 100 Hz -> 10 ms per loop iteration
 
     //main loop for the CLI
-    while (true) {
+    while (running.load()) {
         auto loop_start_time = std::chrono::steady_clock::now();
 
+        std::string input;
         std::cout << "> ";
         std::getline(std::cin, input);
 
         if (input == "exit") {
-            break;
+            running.store(false);
         } else {
-            handleCommand(input);
+            if (input == "0" || input == "pilot") {
```

Comment:
Just wondering why you took this out of the cli.cpp file? I figure having the stuff relating to CLI in the cli.cpp file would mean better "abstraction" (in not an OOP way but still). I'm not opposed to this I'm just wondering

Reply:
The main reason I did that stems from a bug I was finding in `gui.cpp`. Essentially, I wanted to create a ros service client outside of the gui thread (so that you can request camera URLs in CLI). However, the lines to initialize ROS, `rclcpp::init(0, nullptr);`, were in the GUI thread. 

Removing them from the GUI thread and adding them to `main.cpp` caused the program to crash/freeze when you close the GUI because the ROS node inside of it wasn't being destroyed (we were getting stuck at `ros_spin_thread.join()`). I managed to fix it but I noticed that it was stacking instances of the controller input publisher whenever you open and close the GUI. At the end I decided to make one publisher in `main.cpp` and just pass it into whatever threads that wanted it. 

Since we previously used to create the GUI thread in `handleCommand`, I didn't want to increase the number of arguments to that function by adding the controller publisher and whatever subsequent publisher/subscriber nodes we decide to add to the GUI. Therefore, I just put it's contents in `main` where all that stuff was already defined.